src/tls_stream.rs

```diff
@@ -195,6 +249,10 @@ pub struct TlsStream<S> {
 pub enum HandshakeError<S> {
     /// A fatal I/O error occurred
     Failure(io::Error),
+    
+    // Fatal I/O error, with security context.
+    Failure2(io::Error, SecurityContext),
```

Comment:
Adding this variant breaks backwards compatibility - we'll need to make a new enum.